### PR TITLE
speed up X86_insn_reg_intel by using balanced bst

### LINKED ISSUE #719: X86_insn_reg_intel is incredibly slow
in profiling radare2, i noticed a lot of time is spent in the function `X86_insn_reg_intel`, looping over a 115 item array when it could be replaced with a simple lut.

### PR DESCRIPTION
**Pull request by crowell**
should be much faster, with a maximum of 7 comparisons vs 155.

fixes #719

### CODE REVIEW

**Comment by crowell** [comment]
!img

much faster now!

**Comment by Maijin** [comment]
Awesome when merged ?

File: arch/X86/X86Mapping.c [hunk 1]
```diff
@@ -2751,21 +2751,66 @@ static struct insn_reg2 insn_regs_intel2[] = {
 	{ X86_INVLPGA64, X86_REG_RAX, X86_REG_ECX, CS_AC_READ, CS_AC_READ },
 };
 
+struct insn_reg_node_t* new_insn_reg_node(struct insn_reg* data) {
```
Comment (aquynh) [inline arch/X86/X86Mapping.c]: please put the open bracket "{" of function at the beginning of the new line.

File: arch/X86/X86Mapping.h [hunk 2]
```diff
@@ -71,4 +71,16 @@ void X86_reg_access(const cs_insn *insn,
...
+       uint16_t insn;
+       struct insn_reg* value;
+       struct insn_reg_node_t* left;
+       struct insn_reg_node_t* right;
+};
+
+struct insn_reg_node_t* new_insn_reg_node(struct insn_reg* data);
+struct insn_reg_node_t* regs_to_bst(struct insn_reg* arr, int start, int end);
+struct insn_reg* insn_id_to_reg(struct insn_reg_node_t* bst, unsigned int id);
+//void free_bst (struct insn_reg_node_t* node);
```
Comment (aquynh) [inline arch/X86/X86Mapping.h]: please remove this line

File: arch/AArch64/AArch64InstPrinter.c [hunk 3]
```diff
@@ -83,7 +83,7 @@ static void set_mem_access(MCInst *MI, bool status)
 	}
 }
 
-void AArch64_printInst(MCInst *MI, SStream *O, void *Info)
+void AArch64_printInst(struct cs_struct* cs, MCInst *MI, SStream *O, void *Info)
```
Comment (aquynh) [inline arch/AArch64/AArch64InstPrinter.c]: please put * right before cs, like:

    void AArch64_printInst(struct cs_struct *cs, MCInst *MI, SStream *O, void *Info)

File: utils.h [hunk 4]
```diff
@@ -62,6 +62,6 @@ int cs_snprintf(char *buffer, size_t size, const char *fmt, ...);
 bool arr_exist8(unsigned char *arr, unsigned char max, unsigned int id);
 
 bool arr_exist(uint16_t *arr, unsigned char max, unsigned int id);
-
+void free_bst (struct insn_reg_node_t* node);
```
Comment (aquynh) [inline utils.h]: please remove the space between function name and "("

**Comment by danghvu** [comment]
Is it possible to just sort `insn_regs_intel` by id (or create an auxilary sorted array) ? Then you can do a binary search instead of creating a tree structure, searching in an array is faster than following pointers.

**Comment by crowell** [comment]
let me check, that should be more simple.

On Tue, Jun 28, 2016 at 10:56 AM, Hoang-Vu Dang <notifications@github.com>
wrote:

> Is it possible to just sort insn_regs_intel by id (or create an auxilary
> sorted array) ? Then you can do a binary search instead of creating a tree
> structure, searching in an array is faster than following pointers.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/aquynh/capstone/pull/722#issuecomment-229075136>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAlnYF5wD__i4YM_32tDX5mbpdneaGiOks5qQTYegaJpZM4I_po6>
> .
>



-- 
Jeff Crowell
https://github.com/crowell

**Comment by crowell** [comment]
@danghvu see https://github.com/aquynh/capstone/pull/723 for the alternate impl.

**Pull request closed without merging**